js/widgets/musickeyboard.js

```diff
@@ -2638,187 +2632,213 @@ function MusicKeyboard(activity) {
-                } else {
-                    // the previous note block
-                    newStack[endOfStackIdx][4][n - 1] = idx;
+            const newNotes = this._clusterNotes(currentSelectedNotes);
+            const newStack = [
+                [0, ["action", { collapsed: false }], 100, 100, [null, 1, 2, null]],
+                [1, ["text", { value: _("action") + "" + actionGroup }], 0, 0, [0]],
+                [2, "hidden", 0, 0, [0, selectedNotes.length == 0 ? null : 3]]
+            ];
+
+            let prevId = 2;
+            let endOfStackIdx, id;
+
+            for (let noteGrp = 0; noteGrp < newNotes.length; noteGrp++) {
+                const selectedNotesGrp = newNotes[noteGrp];
+                const isLast = noteGrp == newNotes.length - 1;
+                id = newStack.length;
+                let voice = selectedNotes[selectedNotesGrp[0]].voice[0] || DEFAULTVOICE;
+                // Don't use a drum name with set timbre.
+                if (selectedNotes[selectedNotesGrp[0]].noteOctave[0] === "drumnull") {
+                    voice = DEFAULTVOICE;
                 }
+                const next = isLast ? null : id + 3 + this.findLen(selectedNotesGrp, selectedNotes);
 
-                endOfStackIdx = idx;
-
-                const delta = 5;
-
-                // Add a vspace to prevent divide block from obscuring the pitch block.
-                newStack.push([idx + 1, "vspace", 0, 0, [idx, idx + delta]]);
-
-                // note value is saved as a fraction
-                newStack.push([idx + 2, "divide", 0, 0, [idx, idx + 3, idx + 4]]);
-                const maxWidth = Math.max.apply(Math, note.duration);
-
-                const obj = toFraction(maxWidth);
-                newStack.push([idx + 3, ["number", { value: obj[0] }], 0, 0, [idx + 2]]);
-                newStack.push([idx + 4, ["number", { value: obj[1] }], 0, 0, [idx + 2]]);
+                newStack.push(
+                    [id, "settimbre", 0, 0, [prevId, id + 1, id + 3, id + 2]],
+                    [id + 1, ["voicename", { value: voice }], 0, 0, [id]],
+                    [id + 2, "hidden", 0, 0, [id, next]]
+                );
 
-                let thisBlock = idx + delta;
+                prevId = id + 2;
+                endOfStackIdx = id;
 
-                // We need to point to the previous note or pitch block.
-                let previousBlock = idx + 1; // Note block
+                for (let i = 0; i < selectedNotesGrp.length; i++) {
+                    const note = selectedNotes[selectedNotesGrp[i]];
 
-                // The last connection in last pitch block is null.
-                const lastConnection = null;
+                    // Add the Note block and its value
+                    const idx = newStack.length;
+                    newStack.push([idx, "newnote", 0, 0, [endOfStackIdx, idx + 2, idx + 1, null]]);
+                    const n = newStack[idx][4].length;
```

Comment:
What is `n`? can we have any better variable name? thoughts?

Reply:
I didn't make changes to this part of the code, it was already there . I'll have to look at it for more reference on the variable

Reply:
after looking into it , it seems that n refers to the number of connections each note has, 
```
[[0, 'newnote', 100, 100, [null, 1, 4, null]],
 [1, 'divide', 0, 0, [0, 2, 3]],
 [2, ['number', {'value': 1}], 0, 0, [1]],
 [3, ['number', {'value': 4}], 0, 0, [1]],
 [4, 'vspace', 0, 0, [0, 5]],
 [5, 'pitch', 0, 0, [4, 6, 7, null]]
 [6, ['solfege', {'value': 'sol'}], 0, 0, [5]],
 [7, ['number', {'value': 4}], 0, 0, [5]],
]
```
here the length of 4th index refers to the number of connections each note has as far as my understanding goes

Reply:
There is a README file that explains all of this in more detail:

https://github.com/sugarlabs/musicblocks/blob/master/js/README.md#about-the-internal-block-format

Reply:
can you please tell the changes needed exactlly for the pr? 
The code seems to work and the changes you asked before,i.e, removing extra lines and missing spaces, i changed it using prettier